src/StorageRegistry.sol

```diff
@@ -0,0 +1,321 @@
+     *
+     * @param fid   The fid that will receive the storage allocation.
+     * @param units Number of storage units to rent.
+     */
+    function rent(uint256 fid, uint256 units) external payable whenNotDeprecated {
+        if (msg.value != price(units)) revert InvalidPayment();
+        if (rentedUnits + units > maxUnits) revert ExceedsCapacity();
+
+        rentedUnits += units;
+        emit Rent(msg.sender, fid, units);
+    }
+
+    /**
+     * @notice Rent storage for multiple fids. The caller must provide an exact payment amount equal to
+     *         the sum of the prices for each fid's storage allocation.
+     *
+     * @param fids  An array of fids.
+     * @param units An array of storage unit quantities. Must be the same length as the fids array.
+     */
+    function batchRent(uint256[] calldata fids, uint256[] calldata units) external payable whenNotDeprecated {
+        if (fids.length == 0 || units.length == 0) revert InvalidBatchInput();
+        if (fids.length != units.length) revert InvalidBatchInput();
+
+        uint256 _ethUsdPrice = ethUsdPrice();
+        uint256 _usdUnitPrice = usdUnitPrice;
+
+        uint256 totalCost;
+        for (uint256 i; i < fids.length; ++i) {
+            uint256 qty = units[i];
+            if (qty == 0) continue;
+            if (rentedUnits + qty > maxUnits) revert ExceedsCapacity();
+            totalCost += _price(qty, _usdUnitPrice, _ethUsdPrice);
+            rentedUnits += qty;
+            emit Rent(msg.sender, fids[i], qty);
+        }
+
+        if (msg.value != totalCost) revert InvalidPayment();
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                              PRICE VIEWS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Cost in wei to rent one storage unit.
+     */
+    function unitPrice() public view returns (uint256) {
+        return price(1);
+    }
+
+    /**
+     * @notice Calculate the cost in wei to rent the given number of storage units.
+     *
+     * @param units Number of storage units.
+     */
+    function price(uint256 units) public view returns (uint256) {
+        return _price(units, usdUnitPrice, ethUsdPrice());
+    }
+
+    function ethUsdPrice() public view returns (uint256) {
+        /* Ensure that the L2 sequencer is up. */
+        (, int256 sequencerUp, uint256 startedAt,,) = uptimeFeed.latestRoundData();
+        if (sequencerUp != 0) revert SequencerDown();
+
+        /* If the L2 sequencer recently restarted, ensure the grace period has elapsed. */
+        uint256 timeSinceUp = block.timestamp - startedAt;
+        if (timeSinceUp < L2_DOWNTIME_GRACE_PERIOD) revert GracePeriodNotOver();
+
+        /* Get and validate the Chainlink ETH/USD price. */
+        (uint80 roundId, int256 answer,, uint256 updatedAt, uint80 answeredInRound) = priceFeed.latestRoundData();
```

Comment:
Is it possible that `answer` could be set super low or high due to an error upstream? Wondering if we should guard with some sane boundaries.

Reply:
Definitely possible, historically unlikely, there's a first time for everything. I think a good design is to trigger a circuit breaker if the price hits some min/max bound. 

Zooming out, I think the bigger question is what we want the failure mode to be for hubs, clients, and users if storage rentals are down because a dependency (sequencer, price feed) is broken.